### PR TITLE
CommonJS-fix/cache busting

### PR DESCRIPTION
**Pull request by jmcarp**
# Purpose
As described in #1432, names of asset files don't change when their contents do, which means that browsers may inappropriately cache these files. This patch saves assets by content hash and updates the Flask app to include hashes in mako script includes.

# Changes
* Split webpack config into dev and prod files
* Add webpack plugins for minification instead of -p
* Name files by chunk hashes and save hash names to json
* Add `webpack_asset` mako filter for asset includes
* Run webpack in watch mode in dev config file
* Updated webpack invoke task
* Running `webpack` or `inv webpack` will now launch webpack in development mode
    - The equivalent of running `webpack --development --watch`
* Don't minify `conference.js`; minifying breaks Slickgrid filters

# Side effects
* Production webpack config no longer uses the `--production` option as the minification options are now specified
* In prod mode, must restart app server after the complete repack for asset changes to take effect

# To deploy:
* `npm install`
* for dev mode
    * `invoke webpack`
* for prod mode
    * `invoke webpack --production`

# Notes
* This patch doesn't cache-bust `vendor.bundle.js`. Once this is reviewed by @sloria, we can discuss a fix for this.
* No hashes are added to the files in develop mode

### CODE REVIEW

**Comment by sloria** [comment]
To prevent errors in deployment, maybe we should do production mode by default, i.e. `inv webpack` does minification, etc.

File: webpack.dist.config.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
+var webpack = require('webpack');
+var path = require('path');
+var root = path.join(__dirname, 'website', 'static');
+var entry = require('./webpack.common.config.js');
+/** Return the absolute path given a path relative to ./website/static */
+var staticPath = function(dir) {
+    return path.join(root, dir);
+};
+
+SaveAssetsJson = require('assets-webpack-plugin');
```
Comment (sloria) [inline webpack.dist.config.js]: This variable should is undeclared (i.e. missing `var`).

**Comment by sloria** [comment]
We should minimize duplication between the prod and dev config files. Perhaps we could do merge a "common" config object with the dev and prod config objects?

**Comment by jmcarp** [comment]
We actually made a `webpack.common.config.js` file and forgot to git add it. Fixing now.

File: package.json [hunk 2]
```diff
@@ -7,5 +7,8 @@
     "exports-loader": "^0.6.2",
     "imports-loader": "^0.6.3",
     "webpack": "^1.4.13"
+  },
+  "dependencies": {
+    "assets-webpack-plugin": "^0.1.0"
```
Comment (sloria) [inline package.json]: Perhaps we should put `webpack` and all the loaders in `"dependencies"`.

**Comment by sloria** [comment]
Rename `webpack.dist.config.js` to `webpack.prod.config.js`? We usually use "dist" for config files that are meant to be copied but not used, e.g. `local-dist.py`.

File: tasks.py [hunk 3]
```diff
@@ -661,17 +661,19 @@ def clean_assets():
 
 
 @task(aliases=['pack'])
-def webpack(clean=False, watch=False, production=False):
+def webpack(clean=False, watch=False, develop=False):
```
Comment (sloria) [inline tasks.py]: The `assets` task will need to be updated to use the `develop` param.

**Comment by sloria** [comment]
Re: cache-busting `vendor.bundle.js`: The `CommonChunksPlugin` allows you to put `[hash]` in the `filename` argument: https://webpack.github.io/docs/list-of-plugins.html#commonschunkplugin . So I think using:

```js
    new webpack.optimize.CommonsChunkPlugin('vendor', 'vendor.[hash].js'),
```

in the production config file should do the trick.

**Comment by jmcarp** [comment]
@sloria The problem is that the `assets-webpack-plugin` plugin doesn't capture the hash of the vendor file--otherwise we would've done this. I'm sure there's some workaround. Will look into this.

**Comment by sloria** [comment]
Is there any reason not to do cache-busting in develop mode? Does the hashing slow down the build process a lot?

**Comment by sloria** [comment]
Getting a JS error in `vendor.bundle.js` when loading any page after building with the production config:

```
 Uncaught TypeError: Cannot read property 'call' of undefined
```

File: website/util/__init__.py [hunk 4]
```diff
@@ -68,3 +73,23 @@ def is_json_request():
...
+    try:
+        return json.load(open('webpack-assets.json'))
+    except IOError:
+        logger.error('No "webpack-assets.json" file found. You may need to run webpack.')
+        raise
+
+
+asset_paths = load_asset_paths()
+base_static_path = '/static/public/js/'
+def webpack_asset(path):
```
Comment (sloria) [inline website/util/__init__.py]: Could you please add a docstring to this function, indicating that it is meant to be used as a mako filter?

**Comment by sloria** [comment]
1st pass finished. +1 to this approach.

File: website/settings/defaults.py [hunk 5]
```diff
@@ -48,7 +51,7 @@ def parent_dir(path):
 SECRET_KEY = 'CHANGEME'
 
 # May set these to True in local.py for development
-DEV_MODE = False
+DEV_MODE = True
```
Comment (sloria) [inline website/settings/defaults.py]: Why was this changed?

File: website/settings/defaults.py [hunk 6]
```diff
@@ -22,6 +24,7 @@ def parent_dir(path):
 ROOT = os.path.join(BASE_PATH, '..')
 
 LOAD_BALANCER = False
+PROXY_ADDRS = []
```
Comment (sloria) [inline website/settings/defaults.py]: For waterbutler?

File: webpack.prod.config.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+var SaveAssetsJson = require('assets-webpack-plugin');
+
+module.exports = {
+    // Split code chunks by page
+    entry: common.entry,
+    resolve: common.resolve,
+    externals: common.externals,
+    output: common.output,
+    debug: true,
```
Comment (sloria) [inline webpack.prod.config.js]: This should be `false`.

**Comment by sloria** [comment]
For consistency, maybe rename `webpack.config.js` to `webpack.dev.config.js`.

File: website/util/__init__.py (hunk 4 above)
Comment (sloria) [inline website/util/__init__.py]: Could you add some basic unit tests in `tests/test_utils.py` for this function?

**Comment by sloria** [comment]
2nd pass finished. :v:
